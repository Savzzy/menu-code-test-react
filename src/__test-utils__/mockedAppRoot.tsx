import { render, RenderResult } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import { ThemeProvider } from "styled-components";
import globalTheme from "../globalTheme";
import rootReducer from "../reducers";
import { intialStore } from "../store";
import { Store } from "../types";

interface MockedAppRootOptions {
  mockStore: Store;
}

const mockedAppRoot = (
  ComponentTree: JSX.Element,
  options?: MockedAppRootOptions,
): RenderResult => {
  const AppContainer: React.FC = ({ children }) => (
    <Provider
      store={createStore(rootReducer, options?.mockStore || intialStore)}
    >
      <ThemeProvider theme={globalTheme}>{children}</ThemeProvider>
    </Provider>
  );

  return render(ComponentTree, { wrapper: AppContainer });
};

export default mockedAppRoot;
