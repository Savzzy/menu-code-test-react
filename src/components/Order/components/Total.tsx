import React from "react";
import { useSelector } from "react-redux";
import styled from "styled-components";
import { Order, OrderedItemType, Store } from "../../../types";

const TotalContainer = styled.div`
  border-color: ${(props) => props.theme.colors.seperator};
  border-width: 1px;
  border-top-style: solid;
  border-bottom-style: solid;

  display: flex;
  flex-direction: column;
  justify-content: space-between;
  margin-top: 25px;
  padding: 15px 5px;
`;

const PriceRow = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  width: 100%;
  margin: 10px 0;
`;

const DinerTotalLabel = styled.div`
  font-size: 18px;
  color: ${(props) => props.theme.colors.secondaryText};
  font-weight: 400;
  text-transform: capitalize;
`;

const DinerTotalPrice = styled.div`
  font-size: 20px;
  color: ${(props) => props.theme.colors.secondaryText};
  font-weight: 400;
`;

const TotalLabel = styled.div`
  font-size: 20px;
  color: ${(props) => props.theme.colors.primaryText};
  font-weight: 400;
`;

const OrderPrice = styled.div`
  font-size: 25px;
  color: ${(props) => props.theme.colors.primaryText};
  font-weight: 700;
`;

interface DinerTotalType {
  [key: string]: number;
}

const Total: React.FC = () => {
  const orderedItems = useSelector((store: Store) => {
    return store.order;
  });

  const calculateTotal = (
    order: Order,
  ): { dinerTotals: DinerTotalType; orderTotal: number } => {
    const dinerTotals = {};
    let orderTotal = 0;

    Object.keys(order).forEach((key: string) => {
      dinerTotals[key] = 0;

      order[key].orderedItems.forEach((item: OrderedItemType) => {
        dinerTotals[key] += item.price;
      });

      orderTotal += dinerTotals[key];
    });

    return {
      dinerTotals,
      orderTotal,
    };
  };

  const { dinerTotals, orderTotal } = calculateTotal(orderedItems);

  if (orderTotal === 0) {
    return null;
  }

  return (
    <TotalContainer>
      {Object.keys(dinerTotals).map((diner: string) => {
        const dinerTotal = dinerTotals[diner];

        if (dinerTotal === 0) {
          return null;
        }

        return (
          <PriceRow key={diner}>
            <DinerTotalLabel>{diner}</DinerTotalLabel>
            <DinerTotalPrice>{`£ ${dinerTotal.toFixed(2)}`}</DinerTotalPrice>
          </PriceRow>
        );
      })}
      <PriceRow>
        <TotalLabel>Order Total</TotalLabel>
        <OrderPrice>{`£ ${orderTotal.toFixed(2)}`}</OrderPrice>
      </PriceRow>
    </TotalContainer>
  );
};

export default Total;
