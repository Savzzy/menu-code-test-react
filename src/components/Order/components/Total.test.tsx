import React from "react";
import Total from "./Total";
import mockedAppRoot from "../../../__test-utils__/mockedAppRoot";
import { useSelector } from "react-redux";
import {
  mockStore,
  BillAmount,
  mockStoreForNoOrders,
} from "../__meta__/Order.fixtures";

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn(),
}));

describe("Total", () => {
  beforeEach(() => {
    (useSelector as jest.Mock).mockImplementation((getStoreValues) => {
      return getStoreValues(mockStore);
    });
  });

  afterEach(() => {
    (useSelector as jest.Mock).mockClear();
  });

  it("rendered correctly", () => {
    const { container } = mockedAppRoot(<Total />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renderes total bill amount correctly when both diner orders are passed", () => {
    const { getByText } = mockedAppRoot(<Total />);

    expect(getByText(BillAmount)).toBeInTheDocument();
  });
});

describe("Total", () => {
  it("doessn't doesn't render itelf if no orders placed", () => {
    (useSelector as jest.Mock).mockImplementation((getStoreValues) => {
      return getStoreValues(mockStoreForNoOrders);
    });

    const { queryByText } = mockedAppRoot(<Total />);

    expect(queryByText("Order Total")).not.toBeInTheDocument();
  });

  (useSelector as jest.Mock).mockClear();
});
