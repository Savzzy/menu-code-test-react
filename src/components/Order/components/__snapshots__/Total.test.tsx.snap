// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Total rendered correctly 1`] = `
<div
  class="sc-AxjAm eWtYSS"
>
  <div
    class="sc-AxirZ kSgcWt"
  >
    <div
      class="sc-AxiKw goNfIz"
    >
      diner 1
    </div>
    <div
      class="sc-AxhCb daIOJP"
    >
      £ 18.00
    </div>
  </div>
  <div
    class="sc-AxirZ kSgcWt"
  >
    <div
      class="sc-AxiKw goNfIz"
    >
      diner 2
    </div>
    <div
      class="sc-AxhCb daIOJP"
    >
      £ 17.00
    </div>
  </div>
  <div
    class="sc-AxirZ kSgcWt"
  >
    <div
      class="sc-AxhUy hCTVPE"
    >
      Order Total
    </div>
    <div
      class="sc-AxgMl cPaElk"
    >
      £ 35.00
    </div>
  </div>
</div>
`;
